drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,574 @@
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	uint32_t pair = channel;
+	uint32_t channel_mask = (1 << channel);
+	uint32_t unit_time_nsec = (1000000000U / data->cycles_per_sec);
+
+	LOG_DBG("=== Enter %s ...\n", __func__);
+
+	if (!data->capture[pair].callback) {
+		LOG_ERR("PWM capture not configured");
+		return -EINVAL;
+	}
+
+	if (data->capture[pair].is_busy == true) {
+		LOG_ERR("Capture already active on this channel %d", pair);
+		return -EBUSY;
+	}
+
+	data->capture[pair].is_busy = true;
+
+	/* Set capture configuration */
+	EPWM_ConfigCaptureChannel(epwm, channel, unit_time_nsec, 0);
+
+	/* Enable Capture Function for EPWM */
+	EPWM_EnableCapture(epwm, channel_mask);
+
+	/* Enable Timer for EPWM */
+	EPWM_Start(epwm, channel_mask);
+
+	EPWM_ClearCaptureIntFlag(epwm, channel,
+				 EPWM_CAPTURE_INT_FALLING_LATCH | EPWM_CAPTURE_INT_RISING_LATCH);
+
+	/* EnableInterrupt */
+	EPWM_EnableCaptureInt(epwm, channel, data->capture[pair].curr_edge_mode);
+
+	LOG_DBG("===[%s] Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x", __func__, channel, epwm->CAPIEN,
+		epwm->CAPIF);
+
+	return 0;
+}
+
+static int pwm_numaker_disable_capture(const struct device *dev, uint32_t channel)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	uint32_t channel_mask = (1 << channel);
+
+	LOG_DBG("=== Enter %s ...\n", __func__);
+
+	data->capture[channel].is_busy = false;
+	EPWM_Stop(epwm, channel_mask);
+	EPWM_ForceStop(epwm, channel_mask);
+	EPWM_DisableCapture(epwm, channel_mask);
+	EPWM_DisableCaptureInt(epwm, channel,
+			       EPWM_CAPTURE_INT_RISING_LATCH | EPWM_CAPTURE_INT_FALLING_LATCH);
+	EPWM_ClearCaptureIntFlag(epwm, channel,
+				 EPWM_CAPTURE_INT_FALLING_LATCH | EPWM_CAPTURE_INT_RISING_LATCH);
+	LOG_DBG("=== Enter %s ...CAPIEN=0x%x\n", __func__, epwm->CAPIEN);
+	return 0;
+}
+
+/*-----------------------------------------------------------------------------------
+ * Get capture cycles between current channel edge until next chnannel edge.
+ * The capture period counter down count from 0x10000, and auto-reload to 0x10000
+ *-----------------------------------------------------------------------------------
+ */
+static int __pwm_numaker_get_cap_cycle(EPWM_T *epwm, uint32_t channel, uint32_t curr_edge,
+				       uint32_t nextEdge, uint32_t *cycles)
+{
+	uint16_t curr_cnt;
+	uint16_t next_cnt;
+	uint32_t next_if_mask;
+	uint32_t capif_base;
+	uint32_t time_out_cnt;
+	int status = 0;
+	uint32_t period_reloads = 0;
+
+	/* PWM counter is timing critical, to avoid print msg from irq_isr until getting cycles */
+	LOG_DBG("=== Enter %s ...\n", __func__);
+
+	EPWM_ClearPeriodIntFlag(epwm, channel);
+
+	capif_base = (nextEdge == EPWM_CAPTURE_INT_FALLING_LATCH) ? EPWM_CAPIF_CFLIF0_Pos
+								  : EPWM_CAPIF_CRLIF0_Pos;
+	next_if_mask = (0x1ul << (capif_base + channel));
+	time_out_cnt = DT_PROP(DT_NODELABEL(sysclk), clock_frequency) / 2; /* 500 ms time-out */
```

Comment:
You may want to put `DT_PROP(DT_NODELABEL(sysclk), clock_frequency)` in a macro so it has a name